### ISSUE TITLE
False Positive: Intentional uses of non-short-circuit boolean operators are still treated as bugs.

### ISSUE THREAD

**Issue by Carlson-JLQ** [google/error-prone#5643]
# False Positive: Intentional uses of non-short-circuit boolean operators are still treated as bugs.

Version
2.48.0

## Checker
- Checker id: `ShortCircuitBoolean`
- Checker description: This checker detects uses of the non-short-circuiting boolean operators `&` and `|` where the short-circuiting operators `&&` and `||` could be used instead.

## Description of the false positive
`diff/output` marks 2 false positives for `ShortCircuitBoolean`. Across these samples, the checker tends to fire on the outer syntax while missing code-level cues that make the pattern harmless, exempted, or at least much weaker than a real bug. The reports read more like over-eager matching than strong bug findings.

### `NegCase4.java`
Source: `output/codeql/DangerousNonCircuitLogic/src/main/java/scensct/core/neg/NegCase4.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: Bitwise AND with boolean operands and side-effect-free right operand should not be flagged.

```java
// Bitwise AND with boolean operands and side-effect-free right operand should not be flagged.
package scensct.core.neg;

public class NegCase4 {
    public static void main(String[] args) {
        boolean left = true;
        boolean right = false;
        // Scenario 4: Right operand is simple boolean variable (no side effects).
        boolean result = left & right;
    }
}
```

### `NegCase3.java`
Source: `output/errorprone/ShortCircuitBoolean/src/main/java/scensct/core/neg/NegCase3.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: Nested boolean & expression should not be flagged as non-short-circuiting boolean operator.

```java
// Nested boolean & expression should not be flagged as non-short-circuiting boolean operator.
package scensct.core.neg;

public class NegCase3 {
    public static void main(String[] args) {
        boolean p = true;
        boolean q = false;
        boolean r = true;
        boolean nested = ((p & q) ? true : false) & (r ? true : false); // Scenario 3: nested directly as operand of another & // [REPORTED LINE]
        System.out.println(nested);
    }
}
```

## Cause analysis
The rule is documented as follows: This checker detects uses of the non-short-circuiting boolean operators `&` and `|` where the short-circuiting operators `&&` and `||` could be used instead.

 The implementation appears to lean too heavily on surface indicators and not enough on what the code actually does.

That is why these reports read like over-approximation. The code often retains a suspicious signature or control-flow skeleton, but once you read the body, the benchmark's claimed exemption is visible in the source itself.

For `ShortCircuitBoolean`, the practical improvement would be to model those source-level exemptions more explicitly instead of reporting every syntactic near-match.

## References
None known.
